topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,462 @@
+/*
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package blockdev
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/probe"
+	"github.com/skydive-project/skydive/topology"
+	tp "github.com/skydive-project/skydive/topology/probes"
+)
+
+const blockdevGroupName string = "block devices"
+
+// Types returned from lsblk JSON
+const multipathType string = "mpath"
+const lvmType string = "lvm"
+const blockGroupType string = "cluster"
+
+// Types for skydive nodes
+const lvmNodeType string = "blockdevlvm"
+const blockdevNodeType string = "blockdev"
+const leafNodeType string = "blockdevleaf"
+
+const managerType string = "blockdev"
+
+// BlockDevice used for JSON parsing
+type BlockDevice struct {
+	Children []BlockDevice `json:"children"`
+
+	Alignment    int    `json:"alignment"`
+	DiscAln      int    `json:"disc-aln"`
+	DiscGran     string `json:"disc-gran"`
+	DiscMax      string `json:"disc-max"`
+	DiscZero     bool   `json:"disc-zero"`
+	Fsavail      string `json:"fsavail"`
+	Fssize       string `json:"fssize"`
+	Fstype       string `json:"fstype"`
+	FsusePercent string `json:"fsuse%"`
+	Fsused       string `json:"fsused"`
+	Group        string `json:"group"`
+	Hctl         string `json:"hctl"`
+	Hotplug      bool   `json:"hotplug"`
+	Kname        string `json:"kname"`
+	Label        string `json:"label"`
+	LogSec       int    `json:"log-sec"`
+	MajMin       string `json:"maj:min"`
+	MinIo        int    `json:"min-io"`
+	Mode         string `json:"mode"`
+	Model        string `json:"model"`
+	Mountpoint   string `json:"mountpoint"`
+	Name         string `json:"name"`
+	OptIo        int    `json:"opt-io"`
+	Owner        string `json:"owner"`
+	Partflags    string `json:"partflags"`
+	Partlabel    string `json:"partlabel"`
+	Parttype     string `json:"parttype"`
+	Partuuid     string `json:"partuuid"`
+	Path         string `json:"path"`
+	PhySec       int    `json:"hpy-sec"`
+	Pkname       string `json:"pkname"`
+	Pttype       string `json:"pttype"`
+	Ptuuid       string `json:"ptuuid"`
+	Ra           int    `json:"ra"`
```

Comment:
It seems that depending of the lsblk version the output types are not the same :( So you would have to probably maintain mutliple struct. With my version I get a string type for the Ra field
lsblk from util-linux 2.32.1

Reply:
Or just comment `Ra` as it's not used

Reply:
An other option would be to json.Unmarshal to get an interface{}
Then use mapstructure to map it to a BlockDevice struct (we already use mapstructure in a couple of places)

Reply:
It is the "Read Ahead" for the device.  I was testing with 2.33.2 and it is an integer in that version.  I'll look at this more.

Reply:
It looks like lsblk version 2.33 "has been improved to differentiate between numbers,
booleans and strings in JSON output".  See:

https://mirrors.edge.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-ReleaseNotes

I added some code to deal with integers and boolean types possibly encoded as strings.  But, I think it introduces a problem with how skydive looks at the meta data - I'm not sure what the issue is, but currently the blockdev meta data doesn't expand in the UI.  (?)

I'll see if I can make it work.